export const metadata = {
  failures: [
    {
      causedBy: [],
      message: 'Workflow failed',
    }
  ],
  workflowName: 'fetch_sra_to_bam',
  workflowProcessingEvents: [
    {
      cromwellId: 'cromid-bbd63bf',
      description: 'Finished',
      timestamp: '2023-04-20T14:27:29.669Z',
      cromwellVersion: '86-b568fac',
    },
    {
      cromwellId: 'cromid-bbd63bf',
      description: 'PickedUp',
      timestamp: '2023-04-20T14:16:46.886Z',
      cromwellVersion: '86-b568fac',
    },
  ],
  actualWorkflowLanguageVersion: '1.0',
  submittedFiles: {
    workflow:
      'version 1.0\n\nimport "../tasks/tasks_ncbi_tools.wdl" as ncbi_tools\n\nworkflow fetch_sra_to_bam {\n    meta {\n        description: "Retrieve reads from the NCBI Short Read Archive in unaligned BAM format with relevant metadata encoded."\n        author: "Broad Viral Genomics"\n        email:  "viral-ngs@broadinstitute.org"\n        allowNestedInputs: true\n    }\n\n    call ncbi_tools.Fetch_SRA_to_BAM\n\n    output {\n        File   reads_ubam                = Fetch_SRA_to_BAM.reads_ubam\n        String sequencing_center         = Fetch_SRA_to_BAM.sequencing_center\n        String sequencing_platform       = Fetch_SRA_to_BAM.sequencing_platform\n        String sequencing_platform_model = Fetch_SRA_to_BAM.sequencing_platform_model\n        String biosample_accession       = Fetch_SRA_to_BAM.biosample_accession\n        String library_id                = Fetch_SRA_to_BAM.library_id\n        String run_date                  = Fetch_SRA_to_BAM.run_date\n        String sample_collection_date    = Fetch_SRA_to_BAM.sample_collection_date\n        String sample_collected_by       = Fetch_SRA_to_BAM.sample_collected_by\n        String sample_strain             = Fetch_SRA_to_BAM.sample_strain\n        String sample_geo_loc            = Fetch_SRA_to_BAM.sample_geo_loc\n        File   sra_metadata              = Fetch_SRA_to_BAM.sra_metadata\n    }\n}\n',
    root: '',
    options: '{\n  "final_workflow_log_dir": https://test.com/filepath"\n}',
    inputs: '{"fetch_sra_to_bam.Fetch_SRA_to_BAM.SRA_ID":"SRR11954115"}',
    workflowUrl: 'https://raw.githubusercontent.com/broadinstitute/viral-pipelines/v2.1.33.16/pipes/WDL/workflows/fetch_sra_to_bam.wdl',
    labels: '{}',
  },
  calls: {
    'main_workflow.hello_and_goodbye': [
      {
        executionStatus: 'Done',
        shardIndex: -1,
        outputs: {
          goodbye_output: 'Goodbye sub world!',
          hello_output: 'Hello sub world!',
        },
        inputs: {
          hello_and_goodbye_input: 'sub world',
        },
        end: '2016-11-17T14:13:41.117-05:00',
        attempt: 1,
        start: '2016-11-17T14:13:39.236-05:00',
        subWorkflowId: 'a6365f91-c807-465a-9186-a5d3da98fe11',
      },
    ],
    'fetch_sra_to_bam.Fetch_SRA_to_BAM': [
      {
        executionStatus: 'Done',
        stdout: 'http://test.com/stdout',
        commandLine:
          "set -e\n# fetch SRA metadata on this record\nesearch -db sra -q \"SRR11954115\" | efetch -mode json -json > SRA.json\ncp SRA.json \"SRR11954115.json\"\n\n# pull reads from SRA and make a fully annotated BAM -- must succeed\nCENTER=$(jq -r .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SUBMISSION.center_name SRA.json)\nPLATFORM=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM | keys[] as $k | \"\\($k)\"' SRA.json)\nMODEL=$(jq -r \".EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM.$PLATFORM.INSTRUMENT_MODEL\" SRA.json)\nSAMPLE=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.IDENTIFIERS.EXTERNAL_ID|select(.namespace == \"BioSample\")|.content' SRA.json)\nLIBRARY=$(jq -r .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.alias SRA.json)\nRUNDATE=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.RUN_SET.RUN.SRAFiles|if (.SRAFile|type) == \"object\" then .SRAFile.date else [.SRAFile[]|select(.supertype == \"Original\")][0].date end' SRA.json | cut -f 1 -d ' ')\n\nif [ \"$PLATFORM\" = \"OXFORD_NANOPORE\" ]; then\n    # per the SAM/BAM specification\n    SAM_PLATFORM=\"ONT\"\nelse\n    SAM_PLATFORM=\"$PLATFORM\"\nfi\n\nsam-dump --unaligned --header \"SRR11954115\" \\\n    | samtools view -bhS - \\\n    > temp.bam\npicard AddOrReplaceReadGroups \\\n    I=temp.bam \\\n    O=\"SRR11954115.bam\" \\\n    RGID=1 \\\n    RGLB=\"$LIBRARY\" \\\n    RGSM=\"$SAMPLE\" \\\n    RGPL=\"$SAM_PLATFORM\" \\\n    RGPU=\"$LIBRARY\" \\\n    RGPM=\"$MODEL\" \\\n    RGDT=\"$RUNDATE\" \\\n    RGCN=\"$CENTER\" \\\n    VALIDATION_STRINGENCY=SILENT\nrm temp.bam\nsamtools view -H \"SRR11954115.bam\"\n\n# emit numeric WDL outputs\necho $CENTER > OUT_CENTER\necho $PLATFORM > OUT_PLATFORM\necho $SAMPLE > OUT_BIOSAMPLE\necho $LIBRARY > OUT_LIBRARY\necho $RUNDATE > OUT_RUNDATE\nsamtools view -c \"SRR11954115.bam\" | tee OUT_NUM_READS\n\n# pull other metadata from SRA -- allow for silent failures here!\ntouch OUT_MODEL OUT_COLLECTION_DATE OUT_STRAIN OUT_COLLECTED_BY OUT_GEO_LOC\nset +e\njq -r \\\n    .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM.\"$PLATFORM\".INSTRUMENT_MODEL \\\n    SRA.json | tee OUT_MODEL\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"collection_date\" or .TAG==\"collection date\")|.VALUE' \\\n    SRA.json | tee OUT_COLLECTION_DATE\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"strain\")|.VALUE' \\\n    SRA.json | tee OUT_STRAIN\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"collected_by\" or .TAG == \"collecting institution\")|.VALUE' \\\n    SRA.json | tee OUT_COLLECTED_BY\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"geo_loc_name\" or .TAG == \"geographic location (country and/or sea)\")|.VALUE' \\\n    SRA.json | tee OUT_GEO_LOC\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.DESIGN.LIBRARY_DESCRIPTOR.LIBRARY_STRATEGY' \\\n    SRA.json | tee OUT_LIBRARY_STRATEGY\n\nset -e\npython3 << CODE\nimport json\nwith open('SRA.json', 'rt') as inf:\n    meta = json.load(inf)\n# reorganize to look more like a biosample attributes tsv\nbiosample = dict((x['TAG'],x['VALUE']) for x in meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['SAMPLE']['SAMPLE_ATTRIBUTES']['SAMPLE_ATTRIBUTE'])\nbiosample['accession'] = meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['SAMPLE']['IDENTIFIERS']['EXTERNAL_ID']['content']\nbiosample['message'] = 'Successfully loaded'\nbiosample['bioproject_accession'] = meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['STUDY']['IDENTIFIERS']['EXTERNAL_ID']['content']\nbiosample['sample_name'] = biosample['isolate']\nfor k,v in biosample.items():\n    if v == 'not provided':\n        biosample[k] = ''\n\n# British to American conversions (NCBI vs ENA)\nus_to_uk = {\n    'sample_name': 'Sample Name',\n    'isolate': 'Sample Name',\n    'collected_by': 'collecting institution',\n    'collection_date': 'collection date',\n    'geo_loc_name': 'geographic location (country and/or sea)',\n    'host': 'host scientific name',\n}\nfor key_us, key_uk in us_to_uk.items():\n    if not biosample.get(key_us,''):\n        biosample[key_us] = biosample.get(key_uk,'')\n\n# write outputs\nwith open('SRR11954115-biosample_attributes.json', 'wt') as outf:\n    json.dump(biosample, outf)\nCODE",
        shardIndex: -1,
        outputs: {
          library_id: 'SAMN14938612_Next_1_ERCC-42',
          sample_collection_date: '2020-03-10',
          biosample_accession: 'SAMN14938612',
          run_date: '2020-06-08',
          sra_metadata: 'https://www.test.com/metadata',
          library_strategy: 'RNA-Seq',
          num_reads: 405900,
          sequencing_center: 'Broad Institute of Harvard and MIT',
          reads_ubam: 'https://test.com/reads.bam',
          sample_strain: 'not applicable',
          biosample_attributes_json: 'https://test.com/attributes.json',
          sample_geo_loc: 'USA: Massachusetts',
          sequencing_platform: 'ILLUMINA',
          sample_collected_by: 'Massachusetts General Hospital',
          sequencing_platform_model: 'Illumina HiSeq 2500',
        },
        runtimeAttributes: {
          preemptible: 'true',
          disk: '750 GB',
          failOnStderr: 'false',
          disks: 'local-disk 750 LOCAL',
          continueOnReturnCode: '0',
          docker: 'quay.io/broadinstitute/ncbi-tools:2.10.7.10',
          maxRetries: '2',
          cpu: '2',
          memory: '6 GB',
        },
        callCaching: {
          allowResultReuse: false,
          effectiveCallCachingMode: 'CallCachingOff',
        },
        inputs: {
          docker: 'quay.io/broadinstitute/ncbi-tools:2.10.7.10',
          disk_size: 750,
          SRA_ID: 'SRR11954115',
          machine_mem_gb: null,
        },
        returnCode: 0,
        jobId: '56ee000a_06b7b2d4bcdb4598a7df6b6acc60418a',
        backend: 'TES',
        start: '2023-04-20T14:16:48.086Z',
        backendStatus: 'Complete',
        compressedDockerSize: 1339143280,
        end: '2023-04-20T14:27:27.707Z',
        dockerImageUsed: 'quay.io/broadinstitute/ncbi-tools@sha256:c6228528a9fa7d3abd78d40821231ec49c122f8c10bb27ae603540c46fc05797',
        stderr: 'https://test.com/stderr',
        callRoot: 'https://test.com/callRoot',
        attempt: 1,
        executionEvents: [
          {
            startTime: '2023-04-20T14:16:48.086Z',
            description: 'Pending',
            endTime: '2023-04-20T14:16:48.087Z',
          },
          {
            startTime: '2023-04-20T14:16:50.758Z',
            description: 'RunningJob',
            endTime: '2023-04-20T14:27:27.561Z',
          },
          {
            startTime: '2023-04-20T14:27:27.561Z',
            description: 'UpdatingJobStore',
            endTime: '2023-04-20T14:27:27.707Z',
          },
          {
            startTime: '2023-04-20T14:16:48.087Z',
            description: 'RequestingExecutionToken',
            endTime: '2023-04-20T14:16:50.735Z',
          },
          {
            startTime: '2023-04-20T14:16:50.735Z',
            description: 'WaitingForValueStore',
            endTime: '2023-04-20T14:16:50.736Z',
          },
          {
            startTime: '2023-04-20T14:16:50.736Z',
            description: 'PreparingJob',
            endTime: '2023-04-20T14:16:50.758Z',
          },
        ],
      },
      {
        executionStatus: 'Failed',
        stdout: 'stdout',
        commandLine:
          "set -e\n# fetch SRA metadata on this record\nesearch -db sra -q \"SRR11954115\" | efetch -mode json -json > SRA.json\ncp SRA.json \"SRR11954115.json\"\n\n# pull reads from SRA and make a fully annotated BAM -- must succeed\nCENTER=$(jq -r .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SUBMISSION.center_name SRA.json)\nPLATFORM=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM | keys[] as $k | \"\\($k)\"' SRA.json)\nMODEL=$(jq -r \".EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM.$PLATFORM.INSTRUMENT_MODEL\" SRA.json)\nSAMPLE=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.IDENTIFIERS.EXTERNAL_ID|select(.namespace == \"BioSample\")|.content' SRA.json)\nLIBRARY=$(jq -r .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.alias SRA.json)\nRUNDATE=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.RUN_SET.RUN.SRAFiles|if (.SRAFile|type) == \"object\" then .SRAFile.date else [.SRAFile[]|select(.supertype == \"Original\")][0].date end' SRA.json | cut -f 1 -d ' ')\n\nif [ \"$PLATFORM\" = \"OXFORD_NANOPORE\" ]; then\n    # per the SAM/BAM specification\n    SAM_PLATFORM=\"ONT\"\nelse\n    SAM_PLATFORM=\"$PLATFORM\"\nfi\n\nsam-dump --unaligned --header \"SRR11954115\" \\\n    | samtools view -bhS - \\\n    > temp.bam\npicard AddOrReplaceReadGroups \\\n    I=temp.bam \\\n    O=\"SRR11954115.bam\" \\\n    RGID=1 \\\n    RGLB=\"$LIBRARY\" \\\n    RGSM=\"$SAMPLE\" \\\n    RGPL=\"$SAM_PLATFORM\" \\\n    RGPU=\"$LIBRARY\" \\\n    RGPM=\"$MODEL\" \\\n    RGDT=\"$RUNDATE\" \\\n    RGCN=\"$CENTER\" \\\n    VALIDATION_STRINGENCY=SILENT\nrm temp.bam\nsamtools view -H \"SRR11954115.bam\"\n\n# emit numeric WDL outputs\necho $CENTER > OUT_CENTER\necho $PLATFORM > OUT_PLATFORM\necho $SAMPLE > OUT_BIOSAMPLE\necho $LIBRARY > OUT_LIBRARY\necho $RUNDATE > OUT_RUNDATE\nsamtools view -c \"SRR11954115.bam\" | tee OUT_NUM_READS\n\n# pull other metadata from SRA -- allow for silent failures here!\ntouch OUT_MODEL OUT_COLLECTION_DATE OUT_STRAIN OUT_COLLECTED_BY OUT_GEO_LOC\nset +e\njq -r \\\n    .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM.\"$PLATFORM\".INSTRUMENT_MODEL \\\n    SRA.json | tee OUT_MODEL\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"collection_date\" or .TAG==\"collection date\")|.VALUE' \\\n    SRA.json | tee OUT_COLLECTION_DATE\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"strain\")|.VALUE' \\\n    SRA.json | tee OUT_STRAIN\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"collected_by\" or .TAG == \"collecting institution\")|.VALUE' \\\n    SRA.json | tee OUT_COLLECTED_BY\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"geo_loc_name\" or .TAG == \"geographic location (country and/or sea)\")|.VALUE' \\\n    SRA.json | tee OUT_GEO_LOC\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.DESIGN.LIBRARY_DESCRIPTOR.LIBRARY_STRATEGY' \\\n    SRA.json | tee OUT_LIBRARY_STRATEGY\n\nset -e\npython3 << CODE\nimport json\nwith open('SRA.json', 'rt') as inf:\n    meta = json.load(inf)\n# reorganize to look more like a biosample attributes tsv\nbiosample = dict((x['TAG'],x['VALUE']) for x in meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['SAMPLE']['SAMPLE_ATTRIBUTES']['SAMPLE_ATTRIBUTE'])\nbiosample['accession'] = meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['SAMPLE']['IDENTIFIERS']['EXTERNAL_ID']['content']\nbiosample['message'] = 'Successfully loaded'\nbiosample['bioproject_accession'] = meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['STUDY']['IDENTIFIERS']['EXTERNAL_ID']['content']\nbiosample['sample_name'] = biosample['isolate']\nfor k,v in biosample.items():\n    if v == 'not provided':\n        biosample[k] = ''\n\n# British to American conversions (NCBI vs ENA)\nus_to_uk = {\n    'sample_name': 'Sample Name',\n    'isolate': 'Sample Name',\n    'collected_by': 'collecting institution',\n    'collection_date': 'collection date',\n    'geo_loc_name': 'geographic location (country and/or sea)',\n    'host': 'host scientific name',\n}\nfor key_us, key_uk in us_to_uk.items():\n    if not biosample.get(key_us,''):\n        biosample[key_us] = biosample.get(key_uk,'')\n\n# write outputs\nwith open('SRR11954115-biosample_attributes.json', 'wt') as outf:\n    json.dump(biosample, outf)\nCODE",
        shardIndex: -1,
        outputs: {
          library_id: 'SAMN14938612_Next_1_ERCC-42',
          sample_collection_date: '2020-03-10',
          biosample_accession: 'SAMN14938612',
          run_date: '2020-06-08',
          sra_metadata: 'sra_metadata',
          library_strategy: 'RNA-Seq',
          num_reads: 405900,
          sequencing_center: 'Broad Institute of Harvard and MIT',
          reads_ubam: 'reads_ubam',
          sample_strain: 'not applicable',
          biosample_attributes_json: 'biosample_attributes_json',
          sample_geo_loc: 'USA: Massachusetts',
          sequencing_platform: 'ILLUMINA',
          sample_collected_by: 'Massachusetts General Hospital',
          sequencing_platform_model: 'Illumina HiSeq 2500',
        },
        runtimeAttributes: {
          preemptible: 'true',
          disk: '750 GB',
          failOnStderr: 'false',
          disks: 'local-disk 750 LOCAL',
          continueOnReturnCode: '0',
          docker: 'quay.io/broadinstitute/ncbi-tools:2.10.7.10',
          maxRetries: '2',
          cpu: '2',
          memory: '6 GB',
        },
        callCaching: {
          allowResultReuse: false,
          effectiveCallCachingMode: 'CallCachingOff',
        },
        inputs: {
          docker: 'quay.io/broadinstitute/ncbi-tools:2.10.7.10',
          disk_size: 750,
          SRA_ID: 'SRR11954115',
          machine_mem_gb: null,
        },
        returnCode: 0,
        jobId: '56ee000a_06b7b2d4bcdb4598a7df6b6acc60418a',
        backend: 'TES',
        start: '2023-04-20T14:16:48.086Z',
        backendStatus: 'Complete',
        compressedDockerSize: 1339143280,
        end: '2023-04-20T14:27:27.707Z',
        dockerImageUsed: 'quay.io/broadinstitute/ncbi-tools@sha256:c6228528a9fa7d3abd78d40821231ec49c122f8c10bb27ae603540c46fc05797',
        stderr: 'stderr',
        callRoot: 'callRoot',
        attempt: 2,
        executionEvents: [
          {
            startTime: '2023-04-20T14:16:48.086Z',
            description: 'Pending',
            endTime: '2023-04-20T14:16:48.087Z',
          },
          {
            startTime: '2023-04-20T14:16:50.758Z',
            description: 'RunningJob',
            endTime: '2023-04-20T14:27:27.561Z',
          },
          {
            startTime: '2023-04-20T14:27:27.561Z',
            description: 'UpdatingJobStore',
            endTime: '2023-04-20T14:27:27.707Z',
          },
          {
            startTime: '2023-04-20T14:16:48.087Z',
            description: 'RequestingExecutionToken',
            endTime: '2023-04-20T14:16:50.735Z',
          },
          {
            startTime: '2023-04-20T14:16:50.735Z',
            description: 'WaitingForValueStore',
            endTime: '2023-04-20T14:16:50.736Z',
          },
          {
            startTime: '2023-04-20T14:16:50.736Z',
            description: 'PreparingJob',
            endTime: '2023-04-20T14:16:50.758Z',
          },
        ],
      },
    ],
    fetch_REMIX: [
      {
        executionStatus: 'Done',
        stdout: 'stdout',
        commandLine:
          "set -e\n# fetch SRA metadata on this record\nesearch -db sra -q \"SRR11954115\" | efetch -mode json -json > SRA.json\ncp SRA.json \"SRR11954115.json\"\n\n# pull reads from SRA and make a fully annotated BAM -- must succeed\nCENTER=$(jq -r .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SUBMISSION.center_name SRA.json)\nPLATFORM=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM | keys[] as $k | \"\\($k)\"' SRA.json)\nMODEL=$(jq -r \".EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM.$PLATFORM.INSTRUMENT_MODEL\" SRA.json)\nSAMPLE=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.IDENTIFIERS.EXTERNAL_ID|select(.namespace == \"BioSample\")|.content' SRA.json)\nLIBRARY=$(jq -r .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.alias SRA.json)\nRUNDATE=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.RUN_SET.RUN.SRAFiles|if (.SRAFile|type) == \"object\" then .SRAFile.date else [.SRAFile[]|select(.supertype == \"Original\")][0].date end' SRA.json | cut -f 1 -d ' ')\n\nif [ \"$PLATFORM\" = \"OXFORD_NANOPORE\" ]; then\n    # per the SAM/BAM specification\n    SAM_PLATFORM=\"ONT\"\nelse\n    SAM_PLATFORM=\"$PLATFORM\"\nfi\n\nsam-dump --unaligned --header \"SRR11954115\" \\\n    | samtools view -bhS - \\\n    > temp.bam\npicard AddOrReplaceReadGroups \\\n    I=temp.bam \\\n    O=\"SRR11954115.bam\" \\\n    RGID=1 \\\n    RGLB=\"$LIBRARY\" \\\n    RGSM=\"$SAMPLE\" \\\n    RGPL=\"$SAM_PLATFORM\" \\\n    RGPU=\"$LIBRARY\" \\\n    RGPM=\"$MODEL\" \\\n    RGDT=\"$RUNDATE\" \\\n    RGCN=\"$CENTER\" \\\n    VALIDATION_STRINGENCY=SILENT\nrm temp.bam\nsamtools view -H \"SRR11954115.bam\"\n\n# emit numeric WDL outputs\necho $CENTER > OUT_CENTER\necho $PLATFORM > OUT_PLATFORM\necho $SAMPLE > OUT_BIOSAMPLE\necho $LIBRARY > OUT_LIBRARY\necho $RUNDATE > OUT_RUNDATE\nsamtools view -c \"SRR11954115.bam\" | tee OUT_NUM_READS\n\n# pull other metadata from SRA -- allow for silent failures here!\ntouch OUT_MODEL OUT_COLLECTION_DATE OUT_STRAIN OUT_COLLECTED_BY OUT_GEO_LOC\nset +e\njq -r \\\n    .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM.\"$PLATFORM\".INSTRUMENT_MODEL \\\n    SRA.json | tee OUT_MODEL\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"collection_date\" or .TAG==\"collection date\")|.VALUE' \\\n    SRA.json | tee OUT_COLLECTION_DATE\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"strain\")|.VALUE' \\\n    SRA.json | tee OUT_STRAIN\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"collected_by\" or .TAG == \"collecting institution\")|.VALUE' \\\n    SRA.json | tee OUT_COLLECTED_BY\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"geo_loc_name\" or .TAG == \"geographic location (country and/or sea)\")|.VALUE' \\\n    SRA.json | tee OUT_GEO_LOC\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.DESIGN.LIBRARY_DESCRIPTOR.LIBRARY_STRATEGY' \\\n    SRA.json | tee OUT_LIBRARY_STRATEGY\n\nset -e\npython3 << CODE\nimport json\nwith open('SRA.json', 'rt') as inf:\n    meta = json.load(inf)\n# reorganize to look more like a biosample attributes tsv\nbiosample = dict((x['TAG'],x['VALUE']) for x in meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['SAMPLE']['SAMPLE_ATTRIBUTES']['SAMPLE_ATTRIBUTE'])\nbiosample['accession'] = meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['SAMPLE']['IDENTIFIERS']['EXTERNAL_ID']['content']\nbiosample['message'] = 'Successfully loaded'\nbiosample['bioproject_accession'] = meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['STUDY']['IDENTIFIERS']['EXTERNAL_ID']['content']\nbiosample['sample_name'] = biosample['isolate']\nfor k,v in biosample.items():\n    if v == 'not provided':\n        biosample[k] = ''\n\n# British to American conversions (NCBI vs ENA)\nus_to_uk = {\n    'sample_name': 'Sample Name',\n    'isolate': 'Sample Name',\n    'collected_by': 'collecting institution',\n    'collection_date': 'collection date',\n    'geo_loc_name': 'geographic location (country and/or sea)',\n    'host': 'host scientific name',\n}\nfor key_us, key_uk in us_to_uk.items():\n    if not biosample.get(key_us,''):\n        biosample[key_us] = biosample.get(key_uk,'')\n\n# write outputs\nwith open('SRR11954115-biosample_attributes.json', 'wt') as outf:\n    json.dump(biosample, outf)\nCODE",
        shardIndex: -1,
        outputs: {
          library_id: 'SAMN14938612_Next_1_ERCC-42',
          sample_collection_date: '2020-03-10',
          biosample_accession: 'SAMN14938612',
          run_date: '2020-06-08',
          sra_metadata: 'sra_metadata',
          library_strategy: 'RNA-Seq',
          num_reads: 405900,
          sequencing_center: 'Broad Institute of Harvard and MIT',
          reads_ubam: 'reads_ubam',
          sample_strain: 'not applicable',
          biosample_attributes_json: 'biosamples_attributes_json',
          sample_geo_loc: 'USA: Massachusetts',
          sequencing_platform: 'ILLUMINA',
          sample_collected_by: 'Massachusetts General Hospital',
          sequencing_platform_model: 'Illumina HiSeq 2500',
        },
        runtimeAttributes: {
          preemptible: 'true',
          disk: '750 GB',
          failOnStderr: 'false',
          disks: 'local-disk 750 LOCAL',
          continueOnReturnCode: '0',
          docker: 'quay.io/broadinstitute/ncbi-tools:2.10.7.10',
          maxRetries: '2',
          cpu: '2',
          memory: '6 GB',
        },
        callCaching: {
          allowResultReuse: false,
          effectiveCallCachingMode: 'CallCachingOff',
        },
        inputs: {
          docker: 'quay.io/broadinstitute/ncbi-tools:2.10.7.10',
          disk_size: 750,
          SRA_ID: 'SRR11954115',
          machine_mem_gb: null,
        },
        returnCode: 0,
        jobId: '56ee000a_06b7b2d4bcdb4598a7df6b6acc60418a',
        backend: 'TES',
        start: '2023-04-20T14:16:48.086Z',
        backendStatus: 'Complete',
        compressedDockerSize: 1339143280,
        end: '2023-04-20T14:27:27.707Z',
        dockerImageUsed: 'quay.io/broadinstitute/ncbi-tools@sha256:c6228528a9fa7d3abd78d40821231ec49c122f8c10bb27ae603540c46fc05797',
        stderr: 'stderr',
        callRoot: 'callRoot',
        attempt: 1,
        executionEvents: [
          {
            startTime: '2023-04-20T14:16:48.086Z',
            description: 'Pending',
            endTime: '2023-04-20T14:16:48.087Z',
          },
          {
            startTime: '2023-04-20T14:16:50.758Z',
            description: 'RunningJob',
            endTime: '2023-04-20T14:27:27.561Z',
          },
          {
            startTime: '2023-04-20T14:27:27.561Z',
            description: 'UpdatingJobStore',
            endTime: '2023-04-20T14:27:27.707Z',
          },
          {
            startTime: '2023-04-20T14:16:48.087Z',
            description: 'RequestingExecutionToken',
            endTime: '2023-04-20T14:16:50.735Z',
          },
          {
            startTime: '2023-04-20T14:16:50.735Z',
            description: 'WaitingForValueStore',
            endTime: '2023-04-20T14:16:50.736Z',
          },
          {
            startTime: '2023-04-20T14:16:50.736Z',
            description: 'PreparingJob',
            endTime: '2023-04-20T14:16:50.758Z',
          },
        ],
      },
      {
        executionStatus: 'Done',
        stdout: 'stdout',
        commandLine:
          "set -e\n# fetch SRA metadata on this record\nesearch -db sra -q \"SRR11954115\" | efetch -mode json -json > SRA.json\ncp SRA.json \"SRR11954115.json\"\n\n# pull reads from SRA and make a fully annotated BAM -- must succeed\nCENTER=$(jq -r .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SUBMISSION.center_name SRA.json)\nPLATFORM=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM | keys[] as $k | \"\\($k)\"' SRA.json)\nMODEL=$(jq -r \".EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM.$PLATFORM.INSTRUMENT_MODEL\" SRA.json)\nSAMPLE=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.IDENTIFIERS.EXTERNAL_ID|select(.namespace == \"BioSample\")|.content' SRA.json)\nLIBRARY=$(jq -r .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.alias SRA.json)\nRUNDATE=$(jq -r '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.RUN_SET.RUN.SRAFiles|if (.SRAFile|type) == \"object\" then .SRAFile.date else [.SRAFile[]|select(.supertype == \"Original\")][0].date end' SRA.json | cut -f 1 -d ' ')\n\nif [ \"$PLATFORM\" = \"OXFORD_NANOPORE\" ]; then\n    # per the SAM/BAM specification\n    SAM_PLATFORM=\"ONT\"\nelse\n    SAM_PLATFORM=\"$PLATFORM\"\nfi\n\nsam-dump --unaligned --header \"SRR11954115\" \\\n    | samtools view -bhS - \\\n    > temp.bam\npicard AddOrReplaceReadGroups \\\n    I=temp.bam \\\n    O=\"SRR11954115.bam\" \\\n    RGID=1 \\\n    RGLB=\"$LIBRARY\" \\\n    RGSM=\"$SAMPLE\" \\\n    RGPL=\"$SAM_PLATFORM\" \\\n    RGPU=\"$LIBRARY\" \\\n    RGPM=\"$MODEL\" \\\n    RGDT=\"$RUNDATE\" \\\n    RGCN=\"$CENTER\" \\\n    VALIDATION_STRINGENCY=SILENT\nrm temp.bam\nsamtools view -H \"SRR11954115.bam\"\n\n# emit numeric WDL outputs\necho $CENTER > OUT_CENTER\necho $PLATFORM > OUT_PLATFORM\necho $SAMPLE > OUT_BIOSAMPLE\necho $LIBRARY > OUT_LIBRARY\necho $RUNDATE > OUT_RUNDATE\nsamtools view -c \"SRR11954115.bam\" | tee OUT_NUM_READS\n\n# pull other metadata from SRA -- allow for silent failures here!\ntouch OUT_MODEL OUT_COLLECTION_DATE OUT_STRAIN OUT_COLLECTED_BY OUT_GEO_LOC\nset +e\njq -r \\\n    .EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.PLATFORM.\"$PLATFORM\".INSTRUMENT_MODEL \\\n    SRA.json | tee OUT_MODEL\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"collection_date\" or .TAG==\"collection date\")|.VALUE' \\\n    SRA.json | tee OUT_COLLECTION_DATE\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"strain\")|.VALUE' \\\n    SRA.json | tee OUT_STRAIN\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"collected_by\" or .TAG == \"collecting institution\")|.VALUE' \\\n    SRA.json | tee OUT_COLLECTED_BY\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE[]|select(.TAG == \"geo_loc_name\" or .TAG == \"geographic location (country and/or sea)\")|.VALUE' \\\n    SRA.json | tee OUT_GEO_LOC\njq -r \\\n    '.EXPERIMENT_PACKAGE_SET.EXPERIMENT_PACKAGE.EXPERIMENT.DESIGN.LIBRARY_DESCRIPTOR.LIBRARY_STRATEGY' \\\n    SRA.json | tee OUT_LIBRARY_STRATEGY\n\nset -e\npython3 << CODE\nimport json\nwith open('SRA.json', 'rt') as inf:\n    meta = json.load(inf)\n# reorganize to look more like a biosample attributes tsv\nbiosample = dict((x['TAG'],x['VALUE']) for x in meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['SAMPLE']['SAMPLE_ATTRIBUTES']['SAMPLE_ATTRIBUTE'])\nbiosample['accession'] = meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['SAMPLE']['IDENTIFIERS']['EXTERNAL_ID']['content']\nbiosample['message'] = 'Successfully loaded'\nbiosample['bioproject_accession'] = meta['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['STUDY']['IDENTIFIERS']['EXTERNAL_ID']['content']\nbiosample['sample_name'] = biosample['isolate']\nfor k,v in biosample.items():\n    if v == 'not provided':\n        biosample[k] = ''\n\n# British to American conversions (NCBI vs ENA)\nus_to_uk = {\n    'sample_name': 'Sample Name',\n    'isolate': 'Sample Name',\n    'collected_by': 'collecting institution',\n    'collection_date': 'collection date',\n    'geo_loc_name': 'geographic location (country and/or sea)',\n    'host': 'host scientific name',\n}\nfor key_us, key_uk in us_to_uk.items():\n    if not biosample.get(key_us,''):\n        biosample[key_us] = biosample.get(key_uk,'')\n\n# write outputs\nwith open('SRR11954115-biosample_attributes.json', 'wt') as outf:\n    json.dump(biosample, outf)\nCODE",
        shardIndex: -1,
        outputs: {
          library_id: 'SAMN14938612_Next_1_ERCC-42',
          sample_collection_date: '2020-03-10',
          biosample_accession: 'SAMN14938612',
          run_date: '2020-06-08',
          sra_metadata: 'sta_metadata',
          library_strategy: 'RNA-Seq',
          num_reads: 405900,
          sequencing_center: 'Broad Institute of Harvard and MIT',
          reads_ubam: 'reads_ubam',
          sample_strain: 'not applicable',
          biosample_attributes_json: 'biosamples_attribute',
          sample_geo_loc: 'USA: Massachusetts',
          sequencing_platform: 'ILLUMINA',
          sample_collected_by: 'Massachusetts General Hospital',
          sequencing_platform_model: 'Illumina HiSeq 2500',
        },
        runtimeAttributes: {
          preemptible: 'true',
          disk: '750 GB',
          failOnStderr: 'false',
          disks: 'local-disk 750 LOCAL',
          continueOnReturnCode: '0',
          docker: 'quay.io/broadinstitute/ncbi-tools:2.10.7.10',
          maxRetries: '2',
          cpu: '2',
          memory: '6 GB',
        },
        callCaching: {
          allowResultReuse: false,
          effectiveCallCachingMode: 'CallCachingOff',
        },
        inputs: {
          docker: 'quay.io/broadinstitute/ncbi-tools:2.10.7.10',
          disk_size: 750,
          SRA_ID: 'SRR11954115',
          machine_mem_gb: null,
        },
        returnCode: 0,
        jobId: '56ee000a_06b7b2d4bcdb4598a7df6b6acc60418a',
        backend: 'TES',
        start: '2023-04-20T14:16:48.086Z',
        backendStatus: 'Complete',
        compressedDockerSize: 1339143280,
        end: '2023-04-20T14:27:27.707Z',
        dockerImageUsed: 'quay.io/broadinstitute/ncbi-tools@sha256:c6228528a9fa7d3abd78d40821231ec49c122f8c10bb27ae603540c46fc05797',
        stderr: 'stderr',
        callRoot: 'callRoot',
        attempt: 2,
        executionEvents: [
          {
            startTime: '2023-04-20T14:16:48.086Z',
            description: 'Pending',
            endTime: '2023-04-20T14:16:48.087Z',
          },
          {
            startTime: '2023-04-20T14:16:50.758Z',
            description: 'RunningJob',
            endTime: '2023-04-20T14:27:27.561Z',
          },
          {
            startTime: '2023-04-20T14:27:27.561Z',
            description: 'UpdatingJobStore',
            endTime: '2023-04-20T14:27:27.707Z',
          },
          {
            startTime: '2023-04-20T14:16:48.087Z',
            description: 'RequestingExecutionToken',
            endTime: '2023-04-20T14:16:50.735Z',
          },
          {
            startTime: '2023-04-20T14:16:50.735Z',
            description: 'WaitingForValueStore',
            endTime: '2023-04-20T14:16:50.736Z',
          },
          {
            startTime: '2023-04-20T14:16:50.736Z',
            description: 'PreparingJob',
            endTime: '2023-04-20T14:16:50.758Z',
          },
        ],
      },
    ],
  },
  outputs: {
    'fetch_sra_to_bam.sra_metadata': 'fetch.json',
    'fetch_sra_to_bam.reads_ubam': 'reads.bam',
    'fetch_sra_to_bam.biosample_accession': 'SAMN14938612',
    'fetch_sra_to_bam.sample_geo_loc': 'USA: Massachusetts',
    'fetch_sra_to_bam.sample_collection_date': '2020-03-10',
    'fetch_sra_to_bam.sequencing_center': 'Broad Institute of Harvard and MIT',
    'fetch_sra_to_bam.sequencing_platform': 'ILLUMINA',
    'fetch_sra_to_bam.library_id': 'SAMN14938612_Next_1_ERCC-42',
    'fetch_sra_to_bam.run_date': '2020-06-08',
    'fetch_sra_to_bam.sample_collected_by': 'Massachusetts General Hospital',
    'fetch_sra_to_bam.sample_strain': 'not applicable',
    'fetch_sra_to_bam.sequencing_platform_model': 'Illumina HiSeq 2500',
  },
  workflowRoot: 'workflowRoot',
  actualWorkflowLanguage: 'WDL',
  status: 'Succeeded',
  workflowLog: 'workflowLog',
  end: '2023-04-20T14:27:29.668Z',
  start: '2023-04-20T14:16:46.888Z',
  id: '56ee000a-3e05-4f21-b4c7-772627729b02',
  inputs: {
    'fetch_sra_to_bam.Fetch_SRA_to_BAM.docker': 'quay.io/broadinstitute/ncbi-tools:2.10.7.10',
    'fetch_sra_to_bam.Fetch_SRA_to_BAM.machine_mem_gb': null,
    'fetch_sra_to_bam.Fetch_SRA_to_BAM.SRA_ID': 'SRR11954115',
  },
  labels: {
    'cromwell-workflow-id': 'cromwell-56ee000a-3e05-4f21-b4c7-772627729b02',
  },
  submission: '2023-04-20T14:16:17.781Z',
};
